cmd/src/dump.rs

```diff
@@ -29,58 +37,113 @@ pub fn check_and_dump_raft_db<E: RaftEngine>(raftdb_path: &str, engine: &E, env:
-                            state.merge_from_bytes(&value)?;
-                            log_batch.put_raft_state(region_id, &state).unwrap();
-                            count_region += 1;
-                        }
-                        // There is only 2 types of keys in raft.
-                        _ => unreachable!(),
-                    }
-                    if count_entry % 1024 == 0 {
-                        let size = engine.consume(&mut log_batch, true).unwrap();
-                        count_size += size;
-                    }
-                    Ok(true)
-                }
+    let mut threads = vec![];
+    let mut senders = vec![];
+    for _ in 0..thread_num {
+        let raft_engine = engine.clone();
+        let raftdb_engine = origin_engine.clone();
+        let count_size = count_size.clone();
+        let (tx, rx) = mpsc::channel::<Option<u64>>();
+        senders.push(tx);
+        let t = std::thread::spawn(move || {
+            run_worker(rx, raftdb_engine, raft_engine, count_size);
+        });
+        threads.push(t);
+    }
+    info!("Start to scan raft log from raftdb and dump into raft engine");
+    let consumed_time = std::time::Instant::now();
+    // Seek all region id from raftdb and seed them to workers.
+    let mut it = origin_engine.iterator().unwrap();
+    let mut valid = it.seek(SeekKey::Key(keys::REGION_RAFT_MIN_KEY)).unwrap();
+    let mut count_seek = 0;
+    while valid {
+        count_seek += 1;
+        let (key, _) = (it.key(), it.value());
+        match keys::decode_raft_key(key) {
+            Err(_) => continue,
+            Ok((id, _)) => {
+                senders[count_seek % thread_num].send(Some(id)).unwrap();
+                count_region += 1;
+                let next_key = keys::raft_log_prefix(id + 1);
+                valid = it.seek(SeekKey::Key(&next_key)).unwrap();
             }
-        })
-        .unwrap();
-    let size = engine.consume(&mut log_batch, true).unwrap();
+        }
+    }
+    for tx in senders {
+        tx.send(None).unwrap();
+    }
+    info!("Scanned all region id and waiting for dump");
+    for t in threads {
+        t.join().unwrap();
+    }
     engine.sync().unwrap();
-    count_size += size;
     info!(
         "Finished dump, total regions: {}; Total bytes: {}; Consumed time: {:?}",
         count_region,
-        count_size,
-        consume_time.elapsed(),
+        count_size.load(Ordering::Relaxed),
+        consumed_time.elapsed(),
     );
     convert_to_dirty_raftdb(raftdb_path);
     check_and_delete_safely(raftdb_path);
 }
 
+// Worker receives region id and scan the related range.
+fn run_worker(
+    rx: Receiver<Option<u64>>,
+    old_engine: RocksEngine,
+    new_engine: RaftLogEngine,
+    count_size: Arc<AtomicUsize>,
+) {
+    let mut batch = new_engine.log_batch(0);
+    while let Some(id) = rx.recv().unwrap() {
+        let mut entries = Some(vec![]);
```

Comment:
You need to add a counter here to count entries in the current log batch.

Reply:
And it's better to change type of `entries` to `vec`, and use `std::mem::take` later.